Pin Naming Convention Design Document.

malavikasajikumar opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
Adding design document pin-names/pin_names.md for pin naming convention.
<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
None.
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

malavikasajikumar committed:
Adding design document pin-names/pin_names.md for pin naming convention.

0xc0170 commented on docs/design-documents/platform/pin-names/pin_names.md:
```diff
@@ -0,0 +1,198 @@
+
+If only 1 LED/BUTTON is on board, do not define additional LEDs/BUTTONs in BSP. This allows for unavailable LEDs/BUTTONs to be caught as a compile error.
+
+    // Do not do the following
```
this should also provide what it should do first, and then not defining LED or BUTTONS as clones just to have all 1-4  available.

Or rather, the above section should contain a note that each of these should contain an unique pin (so no duplications allowed).

malavikasajikumar replied:
@0xc0170 The current pin naming scheme across various Mbed enabled boards are quite vague and different across different boards. The purpose of this document is to get a starting point on describing the pin naming scheme for future boards (and possibly existing boards). With the differences in pin naming convention, there is a fake sense of portability across various Mbed enabled boards. This document writes down guidelines on how to better achieve that portability in a real sense when defining the pin names that are exposed to customers.

bulislaw replied:
Hm, I would just point out the "anti-paterns" or what not to do rather than give examples.

malavikasajikumar replied:
@bulislaw I have collected lists of board support packages that define these pins differently for the various cases. Providing examples seemed like a more definite way of writing down guidelines.

malavikasajikumar replied:
> this should also provide what it should do first, and then not defining LED or BUTTONS as clones just to have all 1-4 available.
> 
> Or rather, the above section should contain a note that each of these should contain an unique pin (so no duplications allowed).

https://github.com/ARMmbed/mbed-os/pull/12838/commits/63a3c8f59c24f96e2d2f2bf9dec17f4671f86571

MarceloSalazar replied:
I think we all agree with that. We could write examples.
It would be great if we could somehow catch these duplicates. Perhaps we could introduce some kind of checks.

0xc0170 commented on docs/design-documents/platform/pin-names/pin_names.md:
```diff
@@ -0,0 +1,198 @@
+    // Arduino Uno (Rev3) header pin connection naming  
+    // Px_xx relates to the processor pin connected to the Arduino Uno (Rev3) header pin
+    
+    D0 = Px_xx,  // UART RX / GPIO 0  
```
What if some of these are not available? I recall use cases where some were just connected  to VCC or similar. This probably is more connected to below (for instance not always i2c is polluted, it could be a case).

kjbracey replied:
I think you'll have to define them as `NC`, otherwise stuff won't compile - if you've got `TARGET_FF_ARDUINO` on, existing code will need that.

0xc0170 replied:
That is my understanding as well, but should be noted here as well.

malavikasajikumar replied:
@0xc0170 @kjbracey-arm How much flexibility can pins of Arduino form factor have for a board to still be tagged as TARGET_FF_ARDUINO? If certain pins are not connected or have multiple functionality, how can code portability be achieved?

bulislaw replied:
But if the for some reason a pin is not defined for the board (whether in Arduino header or just a BUTTONX) does it make sense for the code to compile? I would rather said it should fail rather than pretend to work...

kjbracey replied:
> If certain pins are not connected or have multiple functionality, how can code portability be achieved?

In practice, if the Arduino connector is not fully independent of the rest of the board, as is quite common, then pin conflicts can happen. How much it matters will depend on which pins you're using. Many things wanting to use the connector only want a couple of pins, so good chance of working, even if it's not a 100%-"intact" connector.

Some drivers are capable of surviving with some pins unconnected, and they would be looking for that with `!= NC` tests.

> I would rather said it should fail rather than pretend to work...

Most uses of `NC` lead to either an MBED_ERROR or MBED_ASSERT (I forget which), so it's not subtle. But certainly harder to track than a compile error.

> does it make sense for the code to compile?

It does if the code is capable of being autoconfiguring in some way - having some ability to cope if a pin isn't available. So far that's always been done with `!= NC` tests in the language, which covers both run-time and compile-time (can't do it in the preprocessor, cos pin values are enums). And that's also been used for different layers of "not available". Eg a config file might have explicitly set an `SPI_CS` to `NC`, or it might be `D1` and that in turn might be `NC`. 

And then SPI.cpp  avoids wiggling it if it is `NC`. It uses `DigitalOut::is_connected()` which means "not constructed with `NC`" - the API implies that `NC` something you can see and test for. It's functionally like `nullptr`, really.

Not compiling would be fine if there was some way to `#ifdef` around it, for people who did want to be adaptive.

     #ifdef FF_ARDUINO_HAVE_D3
     DigitalOut cs(D3);
     #else
     MyDummyDigitalOut cs;    // This class would be a good idea, right?
     #endif

or

    #ifdef FF_ARDUINO_HAVE_D3
    SPI spi(D0, D1, D3);
    #else
    SPI spi(D0, D1, NC);
    #endif

Clunky, but doable, if you have the define.

That sort of thing could actually lead to smaller code, by excluding the cs code at compile time. `if (PIN != NC)` directly like that can be compile-time excluded, but more general code like `DigitalOut(PIN)` won't be.

But then it doesn't fit the pattern of having generic `Driver(PinName, PinName)` classes (or the static pinmap equivalent). We treat pin selections as constructor data, not template parameters. We need to be able pass the null pins to the drivers.

With enough template brain-power, I could imagine

    DigitalOut<D3> cs;

and I think D3 not being defined is syntactically legal, and could select a dummy specialisation... But that's a whole other project.

malavikasajikumar replied:
> Most uses of `NC` lead to either an MBED_ERROR or MBED_ASSERT
>
@0xc0170 How can we confirm this?

0xc0170 replied:
Only to browse the implementation. It was a practice as mentioned above. You either allow NC in the target implementation (most of the cases are there in HAL implementation) or disallow (error or assert in this cases). Does that answer the question?

malavikasajikumar replied:
@0xc0170 @kjbracey-arm 
By defining a pin in the PinName enum as NC, the application programs are enabled to test for "no valid connection" (or "not available"). Shouldn't such tests be extended to HAL for all targets to achieve actual code portability? Just wanted to make sure I am understanding this right.

0xc0170 reviewed (commented):
I reviewed  but would like to get more information here - what is this design document trying to achieve? This looks at the moment more as a documentation update - documenting our current pin naming scheme (its mostly about arduino pin maps. It does not touch own extensions.).

jeromecoutant commented on docs/design-documents/platform/pin-names/pin_names.md:
```diff
@@ -0,0 +1,198 @@
+    // Px_xx relates to the processor pin connected to the Arduino Uno (Rev3) header pin
+    
+    D0 = Px_xx,  // UART RX / GPIO 0  
+    D1 = Px_xx,  // UART TX / GPIO 1  
```
I would remove comments for each lines...
Each pin can be more.

malavikasajikumar replied:
@jeromecoutant If each pin is more, isn't that more reason to add comments?

0xc0170 replied:
a comment describing alternative functionality for each pin? It should be stated in the description

jeromecoutant commented on docs/design-documents/platform/pin-names/pin_names.md:
```diff
@@ -0,0 +1,198 @@
+    // ConnectorName is the name of the specific connector you are implementing signals for here  
+    // Px_xx relates to the processor pin connected to the custom/vendor connector pin  
+    
+    ConnectorName_I2C0_SCL  = Px_xx,  // Connector I2C0 SCL  
```
Connector pin is connected to a MCU pin,
so name should be more something like 
ConnectorName_ConnectorPin = Px_xx,

Then you could define some alias:
ConnectorName_I2C_SCL = ConnectorName_ConnectorPin

kjbracey replied:
My reading of that was that this wasn't a function alias (unlike I2C_SCL for Arduino), it was the actual signal name for this connector.

malavikasajikumar replied:
Agree with @kjbracey-arm on this. However, vendors could choose to define connector pins as "ConnectorName_I2C_SCL = ConnectorName_ConnectorPin". Aliasing is okay as long as pin names are defined, for the sake of portability. For example, if multiple boards use the same connector, then ConnectorName_ConnectorPin should be defined, whether as an alias or processor pin.

MarceloSalazar replied:
I'd like us to evaluate the possibility to use namespacing to have clean definition and usage.
What do you think about the following?
- For Arduino connectors: `I2C i2c(ArduinoUno::I2C_SDA , ArduinoUno::I2C_SCL);`
- For MCU pins: `I2C i2c(ADI::MCU::PB_9 , ADI::MCU::PB_8);`

0xc0170 replied:
Using the namespaces 👍  c++1x scoped enum or similar would be useful? But again, as mentioned below, usefulness of preprocessor macro is to check ifdef but same mechanism should be possible with scoped enums?

MarceloSalazar replied:
We need to review implications on using namespaces as it's a C++ specific feature.
The current HAL is based on C - we may want to keep it for compatibility reasons and just use explicit defines for each of the connector standards.

jeromecoutant reviewed (commented):
Do you think about some github action script in order to check pin names ?

kjbracey commented on docs/design-documents/platform/pin-names/pin_names.md:
```diff
@@ -0,0 +1,198 @@
+
+### Arduino Uno (Rev3) Header Pins
+
+All Arduino Uno (Rev3) form factor Mbed controller boards should define D0-D15 & A0-A5 pins as a default standard.
```
They also need to make sure that `"ARDUINO"` is in the `supported-form-factors` section in the target.  That should at least get a mention.

That causes `TARGET__FF_ARDUINO` to be defined, and then lots of existing in-tree code will assume the D0-D15 and A0-A6 pins are present. (Eg drivers that have a default pin selection for Arduino form factor).

ewalsh-adi replied:
(I should say that Malavika, who created this pull request, is a teammate of mine and I'm working with her on this proposal)

Agree.
One further point is it might be an idea to change the form factor name to ARDUINO_UNO and perhaps then TARGET_FF_ARDUINO to TARGET_FF_ARDUINO_UNO. There are Mbed supported Arduino Nano form-factor boards and maybe MKR form-factor as well (if there isn't already then the "Portenta H7" board will change that).

malavikasajikumar replied:
> They also need to make sure that `"ARDUINO"` is in the `supported-form-factors` section in the target. That should at least get a mention.
> 
> That causes `TARGET__FF_ARDUINO` to be defined, and then lots of existing in-tree code will assume the D0-D15 and A0-A6 pins are present. (Eg drivers that have a default pin selection for Arduino form factor).

https://github.com/ARMmbed/mbed-os/pull/12838/commits/63a3c8f59c24f96e2d2f2bf9dec17f4671f86571

malavikasajikumar replied:
> They also need to make sure that `"ARDUINO"` is in the `supported-form-factors` section in the target. That should at least get a mention.
> 
> That causes `TARGET__FF_ARDUINO` to be defined, and then lots of existing in-tree code will assume the D0-D15 and A0-A6 pins are present. (Eg drivers that have a default pin selection for Arduino form factor).

@kjbracey-arm Is it A0-A6 or A0-A5?

MarceloSalazar replied:
I'd suggest that we introduce a new `"ARDUINOUNO"` form factor that follows the proposal being discussed here and we aim to deprecate the existing `"ARDUINO"` when possible. This should help us to ensure that new boards that follow the standard can be tested accordingly and we have a clear definition of pin-names. Let me know your thoughts on this.

kjbracey commented on docs/design-documents/platform/pin-names/pin_names.md:
```diff
@@ -0,0 +1,198 @@
+
+### Avoid Unnecessary Definitions
+
+If only 1 LED/BUTTON is on board, do not define additional LEDs/BUTTONs in BSP. This allows for unavailable LEDs/BUTTONs to be caught as a compile error.
```
Is that what we've done up to now? I don't think so.

The downside of not defining there's no way to conditionally cope if you do that. You can't `#ifdef` an enum value. But if they're defined as NC, you can test for them:

     if (BUTTON2 != NC) {
          // code for button 2
     }

That is effectively a compile-time `if`. In C++17 you could get fancy and force that with `if constexpr (BUTTON2 != NC)`.

If you want a compile error, you can force one with

    static_assert(BUTTON1 != NC && BUTTON2 != NC, "This application needs two buttons");

Otherwise you will get a run-time error when you try to use an NC pin.

0xc0170 replied:
This was one of the problems, static_assert is a good one to be provided here.

jeromecoutant replied:
So why do we keep theses pin name in the enum?
and not use define ?

kjbracey replied:
Legacy convention. You could change it, but currently `#ifdef BUTTON1` is false on all platforms, so tells you nothing. If every platform was universally updated, then that construct would be useful.

Another minor(ish) issue is that defines can be more intrusive. Defining something short like `A0` can cause more collision upset than having an `A0` identifier.

I've seen cases like

    template <typename A0>

which explode if you have

    #define A0 (PIN_H0)   // PIN_H0 is enum value

but you do get away with

    #define A0 PIN_H0

because `template <typename PIN_H0>` still works while `template <typename (PIN_H0)>` doesn't. If `PIN_H0` was `#define`d as something, and not an enum value, then you'd be in trouble.

jeromecoutant replied:
Other solution, is to move alias definition into json config file?

MarceloSalazar replied:
I think it's clear that we need `NA`.
The actual question is what to do with that when the application uses that, then we can figure out implementation details. Options:
- Ignore (silent compile)
- Warning
- Compile error

bulislaw reviewed (commented):
One general comment I have is around the scope of this document. As it is, it addresses the dev boards only. And maybe it is correct, but there's also the production HW and standard pin names (aliases) for the peripherals. Do people think it should live in the HAL spec? In which case we should probably make it obvious that scope of this document is development boards only.

Second option would be to extend this document about general points which will influence the HAL definition for each peripheral.

malavikasajikumar committed:
Adding a few extra notes to address the following comments:
https://github.com/ARMmbed/mbed-os/pull/12838#discussion_r412167373
https://github.com/ARMmbed/mbed-os/pull/12838#discussion_r411974603

MarceloSalazar commented on docs/design-documents/platform/pin-names/pin_names.md:
```diff
@@ -0,0 +1,200 @@
+
+To achieve meaningful portability of application code across various Mbed enabled boards, the pin names of commonly used interfaces and board components should be common across these boards. This document describes a set of rules on how to name these pins in the board support package, specifically in PinNames.h file of the board.
+
+### Arduino Uno (Rev3) Header Pins
```
We should consider the "Arduino Uno" as one of the popular standards and example of actual implementation, but also need to introduce the concept considering that other standards will follow shortly.

malavikasajikumar replied:
> One general comment I have is around the scope of this document. As it is, it addresses the dev boards only. And maybe it is correct, but there's also the production HW and standard pin names (aliases) for the peripherals. Do people think it should live in the HAL spec? In which case we should probably make it obvious that scope of this document is development boards only.
> 
> Second option would be to extend this document about general points which will influence the HAL definition for each peripheral.

@bulislaw The second suggestion here https://github.com/ARMmbed/mbed-os/pull/12838#discussion_r434038918 
> For MCU pins: I2C i2c(ADI::MCU::PB_9 , ADI::MCU::PB_8);
>
can possibly help with easily editing application code for production HW?

MarceloSalazar replied:
I believe we need to define the architecture of the pin names on multiple scenarios.
On dev boards, it could be ArduinoUno or other connectors. On production/custom type of boards, the most logical seems to use MCU names with the possibility to let developers customized for their own needs.

I tend to think we should split the document and proposal into two:
- General architecture and guidelines for pin names (do's and don'ts), including MCUs.
- Arduino Uno specific proposal and guidelines, as the first and reference connector

@bulislaw @0xc0170 does it sound reasonable?

ladislas replied:
agreed with @MarceloSalazar

malavikasajikumar committed:
Adding a first draft of Arduino Uno specific Pin Names design document.
The original proposal has to be divided between an Arduino Uno specific Pin Names design document and a generic Pin Names design document.

malavikasajikumar committed:
Renaming the original Pin Names design document as the Generic Pin Names design document.

malavikasajikumar committed:
Removing Board components section from Arduino Uno specific Pin Names design document. This section should reside in the Generic document.

malavikasajikumar committed:
Removing Arduino Uno specific definition from the generic Pin name design document. Updated board components section.
